Place block storage (performance ISCSI) order by indicating iops, disksize, hourlybillingflag

jianqiu opened the pull request:
This PR is trying to fix the issue [110](https://github.com/maximilien/softlayer-go/issues/110)

maximilien commented on softlayer/softlayer_product_package_service.go:
```diff
@@ -9,6 +9,8 @@ type SoftLayer_Product_Package_Service interface {
 
 	GetItemPrices(packageId int) ([]datatypes.SoftLayer_Product_Item_Price, error)
 	GetItemPricesBySize(packageId int, size int) ([]datatypes.SoftLayer_Product_Item_Price, error)
+	GetItemPricesBySizeAndIops(packageId int, size int, capacity int) ([]datatypes.SoftLayer_Product_Item_Price, error)
```
Hi, @jianqiu, are these service methods in the `SL_Product_Item_Price` service? If yes, then this PR makes sense and should be merged in SL go, if not then let's move these into the SL CPI code. Maybe create them into a file in the `utils` package.  Thoughts?

jianqiu replied:
@maximilien ,
I think it should be a complex service method in the `SoftLayer_Product_Package`. The based method is [getItemPrices](http://sldn.softlayer.com/reference/services/SoftLayer_Product_Package/getItemPrices), what we do is to customize the method with result filter and mask. The following SL blog is the example we followed.
https://softlayer.github.io/python/place_order_block_storage_performance_iscsi/

maximilien commented on test_helpers/mock_softlayer_product_package.go:
```diff
@@ -29,6 +29,14 @@ func (mock *MockProductPackageService) GetItemPricesBySize(packageId int, size i
+	return []datatypes.SoftLayer_Product_Item_Price{}, errors.New("Not supported")
+}
+
+func (mock *MockProductPackageService) GetIopsItemPricesBySize(packageId int, size int) ([]datatypes.SoftLayer_Product_Item_Price, error) {
```
Let's use Fake as we do for all other mocks/fakes. For consistency.

maximilien commented:
Overall looks good @jianqiu, however, see my comments. I think this needs to be in CPI as utility unless the answer to my question is yes and you can point me to SL docs online. Thx

maximilien commented:
Ok @jianqiu what about this proposed compromise:

1. We keep it here as you want since, as you argue, it's an advance use of the SL API
2. We, however, move all non-official SL APIs functions into a new `utils` package

That way we do not break the main design point of SL go to reflect the SL APIs 1-to-1 as much as possible, yet allow useful "extensions" like this one to exist into utilities. We could even in time create things like `utils/storage` and `utils/virtual_guest` etc.

What do you think JQ?

jianqiu commented:
@maximilien ，
Max, I already updated the PR and removed all *special* methods, made just 1-to-1 mapping. And the PR had resolved issue [110](https://github.com/maximilien/softlayer-go/issues/110), Simyung Yang already closed it. Thanks!

maximilien commented on services/fakes/fake_softlayer_product_package.go:
```diff
@@ -0,0 +1,40 @@
+	return productItems, nil
+}
+
+func (fps *FakeProductPackageService) GetItemPrices(packageId int, filters string) ([]datatypes.SoftLayer_Product_Item_Price, error) {
```
Why are we not supporting these methods? Just don't include them in the interface, that's better than returning an error IMO

maximilien commented:
Merging for purposes of solving issue #110 but please address my comment, remove the methods we are not implementing rather than returning some special errors.
